File: bindings/csharp/src/ProvenApiserver.cs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// API Server protocol types for proven-servers.

namespace Proven;

/// <summary>AuthScheme matching the Idris2 ABI tags (0-5).</summary>
public enum AuthScheme : byte
{
    ApiKey = 0,
    Bearer = 1,
    Basic = 2,
    OAuth2 = 3,
    Hmac = 4,
    Mtls = 5
}

/// <summary>RateLimitStrategy matching the Idris2 ABI tags (0-3).</summary>
public enum RateLimitStrategy : byte
{
    FixedWindow = 0,
    SlidingWindow = 1,
    TokenBucket = 2,
    LeakyBucket = 3
}

/// <summary>ApiVersion matching the Idris2 ABI tags (0-4).</summary>
public enum ApiVersion : byte
{
    V1 = 0,
    V2 = 1,
    V3 = 2,
    Latest = 3,
    Deprecated = 4
}

/// <summary>ResponseFormat matching the Idris2 ABI tags (0-3).</summary>
public enum ResponseFormat : byte
{
    Json = 0,
    Xml = 1,
    Protobuf = 2,
    MessagePack = 3
}

/// <summary>GatewayError matching the Idris2 ABI tags (0-5).</summary>
public enum GatewayError : byte
{
    Unauthorized = 0,
    RateLimited = 1,
    NotFound = 2,
    BadRequest = 3,
    ServiceUnavailable = 4,
    CircuitOpen = 5
}

File: bindings/php/src/ProvenApiserver.php
<?php
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// API Server protocol types for proven-servers.

declare(strict_types=1);

namespace ProvenServers;

/** AuthScheme matching the Idris2 ABI tags. */
enum AuthScheme: int
{
    case ApiKey = 0;
    case Bearer = 1;
    case Basic = 2;
    case OAuth2 = 3;
    case Hmac = 4;
    case Mtls = 5;
}

/** RateLimitStrategy matching the Idris2 ABI tags. */
enum RateLimitStrategy: int
{
    case FixedWindow = 0;
    case SlidingWindow = 1;
    case TokenBucket = 2;
    case LeakyBucket = 3;
}

/** ApiVersion matching the Idris2 ABI tags. */
enum ApiVersion: int
{
    case V1 = 0;
    case V2 = 1;
    case V3 = 2;
    case Latest = 3;
    case Deprecated = 4;
}

/** ResponseFormat matching the Idris2 ABI tags. */
enum ResponseFormat: int
{
    case Json = 0;
    case Xml = 1;
    case Protobuf = 2;
    case MessagePack = 3;
}

/** GatewayError matching the Idris2 ABI tags. */
enum GatewayError: int
{
    case Unauthorized = 0;
    case RateLimited = 1;
    case NotFound = 2;
    case BadRequest = 3;
    case ServiceUnavailable = 4;
    case CircuitOpen = 5;
}
